### PR TITLE
Add GCP Workload Identity Federation Auth Strategy

### PR DESCRIPTION
**Pull request by abhishoya-gs**
Add GCP Workload Identity Federation Strategy
Merged Pure PRs: https://github.com/finos/legend-pure/pull/495 , https://github.com/finos/legend-pure/pull/482

### CODE REVIEW

File: .github/workflows/database-bigquery-integration-test.yml [hunk 1]
```diff
@@ -19,6 +19,8 @@ env:
   CI_DEPLOY_PASSWORD: ${{ secrets.CI_DEPLOY_PASSWORD }}
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
Comment (epsstan) [inline .github/workflows/database-bigquery-integration-test.yml:22]: Prefix secrets used in tests with LEGEND_INTEG_TEST

File: legend-engine-executionPlan-execution-store-authentication/src/main/java/org/finos/legend/engine/authentication/flows/BigQueryWithGCPWorkloadIdentityFederationUsingAWSFlow.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    private String getUTCDate(Date date) {
+        SimpleDateFormat dateTimeFormat = new SimpleDateFormat(ISO8601BasicFormat);
+        dateTimeFormat.setTimeZone(new SimpleTimeZone(0, "UTC"));
+        return dateTimeFormat.format(date);
+    }
+
+    private Credentials assumeAWSRoleAndGetCredentials(String roleArn, String roleSessionName, String region, String awsAccessKeyIdVaultReference, String awsSecretAccessKeyVaultReference) {
+        StsClient stsClient = StsClient.builder()
+                .region(Region.of(region))
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(Vault.INSTANCE.getValue(awsAccessKeyIdVaultReference),Vault.INSTANCE.getValue(awsSecretAccessKeyVaultReference))))
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-authentication/src/main/java/org/finos/legend/engine/authentication/flows/BigQueryWithGCPWorkloadIdentityFederationUsingAWSFlow.java:116]: Minor. Indent VAULT api invocation

File: legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationUsingAWSFlow.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.finos.legend.engine.shared.core.identity.factory.IdentityFactoryProvider;
+import org.finos.legend.engine.shared.core.vault.Vault;
+import org.junit.Before;
+import org.junit.Test;
+
+
+import static org.junit.Assert.*;
+
+public class ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationUsingAWSFlow {
+    private final Identity identity1 = IdentityFactoryProvider.getInstance().makeIdentityForTesting("identity1");
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationUsingAWSFlow.java:32]: Replace inmemory vault with env vault 
```
  @BeforeClass
    public static void setupTest() throws IOException
    {
        Vault.INSTANCE.registerImplementation(new EnvironmentVaultImplementation());
    }
```

File: legend-engine-executionPlan-execution-store-relational-connection-tests/pom.xml [hunk 4]
```diff
@@ -182,6 +182,9 @@
                         </dependencies>
                         <!-- Without override, surefire configuration would have skipped our integration tests -->
                         <configuration combine.self="override">
+                            <includes>
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection-tests/pom.xml:185]: Make test selection consistent with other database integration tests

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java [hunk 5]
```diff
@@ -58,12 +58,23 @@ public static void verifyTestSetup()
         {
            // fail(String.format("Tests cannot be run. GCP env variable %s has not been set", GOOGLE_APPLICATION_CREDENTIALS));
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java:59]: Remove GCP comments

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java [hunk 6]
```diff
@@ -58,12 +58,23 @@ public static void verifyTestSetup()
...
         }
+        String awsAccessKeyId = System.getenv(AWS_ACCESS_KEY_ID);
+        if (awsAccessKeyId == null || awsAccessKeyId.trim().isEmpty())
+        {
+            // fail(String.format("Tests cannot be run. AWS env variable %s has not been set", AWS_ACCESS_KEY_ID));
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java:64]: Put back the comments

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java [hunk 7]
```diff
@@ -58,12 +58,23 @@ public static void verifyTestSetup()
...
+        {
+            // fail(String.format("Tests cannot be run. AWS env variable %s has not been set", AWS_SECRET_ACCESS_KEY));
+        }
     }
 
     @BeforeClass
     public static void setupTest() throws IOException
     {
         Vault.INSTANCE.registerImplementation(new EnvironmentVaultImplementation());
+        Vault.INSTANCE.registerImplementation(inMemoryVault);
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java:77]: In memory vault not needed

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java [hunk 8]
```diff
@@ -76,7 +87,10 @@ protected Subject getSubject()
     public void setup()
     {
         LegendDefaultDatabaseAuthenticationFlowProvider flowProvider = new LegendDefaultDatabaseAuthenticationFlowProvider();
+        LegendDefaultDatabaseAuthenticationFlowProviderConfiguration flowProviderConfiguration = new LegendDefaultDatabaseAuthenticationFlowProviderConfiguration();
+        flowProvider.configure(flowProviderConfiguration);
         assertBigQueryFlowIsAvailable(flowProvider);
+        assertBigQueryWifUsingAwsFlowIsAvailable(flowProvider);
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java:93]: Minor. Consistent abbreviations. GCP/WIF vs Gcp/Wif

File: legend-engine-executionPlan-execution-store-authentication/pom.xml [hunk 9]
```diff
@@ -43,8 +43,87 @@
             <groupId>com.fasterxml.jackson.core</groupId>
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-authentication/pom.xml:43]: Figure out how to add Simba to main server classpath

**Review by epsstan — CHANGES_REQUESTED** [review]

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java [hunk 10]
```diff
@@ -56,7 +54,17 @@ public static void verifyTestSetup()
         String googleApplicationCredentials = System.getenv(GOOGLE_APPLICATION_CREDENTIALS);
         if (googleApplicationCredentials == null || googleApplicationCredentials.trim().isEmpty())
         {
-           // fail(String.format("Tests cannot be run. GCP env variable %s has not been set", GOOGLE_APPLICATION_CREDENTIALS));
+//            fail(String.format("Tests cannot be run. GCP env variable %s has not been set", GOOGLE_APPLICATION_CREDENTIALS));
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/ExternalIntegration_TestConnectionAcquisitionWithFlowProvider_BigQuery.java]: Done

File: .github/workflows/database-bigquery-integration-test.yml [hunk 11]
```diff
@@ -19,6 +19,8 @@ env:
   CI_DEPLOY_PASSWORD: ${{ secrets.CI_DEPLOY_PASSWORD }}
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+  AWS_ACCESS_KEY_ID: ${{ secrets.LEGEND_INTEG_TEST_AWS_ACCESS_KEY_ID }}
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply .github/workflows/database-bigquery-integration-test.yml]: Right, including them in steps now

File: legend-engine-language-pure-store-relational/src/main/java/org/finos/legend/engine/language/pure/grammar/from/authentication/AuthenticationStrategyParseTreeWalker.java [hunk 12]
```diff
@@ -87,4 +88,18 @@ public GCPApplicationDefaultCredentialsAuthenticationStrategy visitGCPApplicatio
         authStrategy.sourceInformation = code.getSourceInformation();
         return authStrategy;
     }
+
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-language-pure-store-relational/src/main/java/org/finos/legend/engine/language/pure/grammar/from/authentication/AuthenticationStrategyParseTreeWalker.java:91]: Added [here](https://github.com/finos/legend-engine/pull/589/commits/fb2c4aa1ba052e59980ff5e8174f723f2cefcbd1#diff-a428f06079ed306ac6336a7e6a64b2a7c33a61c068e8983b090f415d9f018901)

File: legend-engine-executionPlan-execution-store-authentication/src/main/java/org/finos/legend/engine/authentication/LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.java [hunk 13]
```diff
@@ -18,8 +18,31 @@
 
 public final class LegendDefaultDatabaseAuthenticationFlowProviderConfiguration extends DatabaseAuthenticationFlowProviderConfiguration
 {
+    public AWSConfig awsConfig;
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-executionPlan-execution-store-authentication/src/main/java/org/finos/legend/engine/authentication/LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.java]: Refactored it [here](https://github.com/finos/legend-engine/pull/589/commits/65bc7a3008c1d4ee308ec60aa945122d7c40985a)

**Review by epsstan — APPROVED** [review]

File: legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationFlow.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    public void makeCredential() throws Exception {
+        LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.AWSConfig awsConfig = new LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.AWSConfig(
+                "us-east-1",
+                "564704738649",
+                "gcp-wif",
+                "AWS_ACCESS_KEY_ID",
+                "AWS_SECRET_ACCESS_KEY"
+        );
+        LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.GCPWorkloadConfig gcpWorkloadConfig = new LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.GCPWorkloadConfig(
+                "412074507462",
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationFlow.java]: Done

File: legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationFlow.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    public void setup()
+    {
+        Vault.INSTANCE.registerImplementation(new EnvironmentVaultImplementation());
+    }
+
+    @Test
+    public void makeCredential() throws Exception {
+        LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.AWSConfig awsConfig = new LegendDefaultDatabaseAuthenticationFlowProviderConfiguration.AWSConfig(
+                "us-east-1",
+                "564704738649",
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/flows/ExternalIntegration_TestBigQueryWithGCPWorkloadIdentityFederationFlow.java]: Done

File: legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/TestDefaultSupportedFlows.java [hunk 16]
```diff
@@ -40,6 +46,7 @@
                 { DatabaseType.BigQuery, BigQueryDatasourceSpecification.class, GCPApplicationDefaultCredentialsAuthenticationStrategy.class },
                 { DatabaseType.SqlServer, StaticDatasourceSpecification.class, UserNamePasswordAuthenticationStrategy.class },
                 { DatabaseType.Databricks, DatabricksDatasourceSpecification.class, ApiTokenAuthenticationStrategy.class },
+                { DatabaseType.BigQuery, BigQueryDatasourceSpecification.class, GCPWorkloadIdentityFederationAuthenticationStrategy.class}
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply legend-engine-executionPlan-execution-store-authentication/src/test/java/org/finos/legend/engine/authentication/TestDefaultSupportedFlows.java]: Done

File: docs/bigquery/bigquery-connector.md [hunk 17]
```diff
@@ -33,6 +33,15 @@ See the following docs for details :
 * https://cloud.google.com/bigquery/docs/authentication
 * https://cloud.google.com/docs/authentication/production 
 
+__Workload Identity Federation__
+
+Legend can connect to BigQuery using Workload Identity Federation. This means that BigQuery JDBC driver will require any valid Service Account Access Token obtained using configured authentication flow during runtime to connect to BigQuery.
```
[the remark this answers is not in the document]
Comment (abhishoya-gs) [reply docs/bigquery/bigquery-connector.md]: Okay, changed

**Pull request merged**